File: cases/taint/csharp/recursive-heap-unwind-negative/RecursiveHeapUnwindNegative.cs
namespace DataFlowBench;

static class RecursiveHeapUnwindNegative
{
    sealed class Box
    {
        public int Value;
    }

    static int dfb_source() // DFB-SOURCE: recursive-heap-unwind-input
    {
        return 7;
    }

    static void dfb_sink(int value) { } // DFB-SINK: recursive-heap-unwind-sink

    static void Walk(Box box, int value, int depth)
    {
        if (depth == 0)
        {
            box.Value = value;
            box.Value = 0; // DFB-KILL: recursive-heap-unwind-base-clean
            return; // DFB-WITNESS: recursive-heap-unwind-base
        }

        Walk(box, value, depth - 1); // DFB-WITNESS: recursive-heap-unwind-transfer
        box.Value = box.Value + 1; // DFB-WITNESS: recursive-heap-unwind-compose
    }

    static void Run()
    {
        Box box = new Box();
        Walk(box, dfb_source(), 3);
        dfb_sink(box.Value);
    }
}

File: cases/taint/scala/recursive-heap-unwind-negative/RecursiveHeapUnwindNegative.scala
package dataflowbench

object RecursiveHeapUnwindNegative {
  class Box(var value: Int = 0)

  def dfb_source(): Int = { // DFB-SOURCE: recursive-heap-unwind-input
    7
  }

  def dfb_sink(value: Int): Unit = {} // DFB-SINK: recursive-heap-unwind-sink

  def walk(box: Box, value: Int, depth: Int): Unit = {
    if (depth == 0) {
      box.value = value // DFB-WITNESS: recursive-heap-unwind-base
      box.value = 0 // DFB-KILL: recursive-heap-unwind-base-clean
    } else {
      walk(box, value, depth - 1) // DFB-WITNESS: recursive-heap-unwind-transfer
      box.value = box.value + 1 // DFB-WITNESS: recursive-heap-unwind-compose
    }
  }

  def run(): Unit = {
    val box = new Box()
    walk(box, dfb_source(), 3)
    dfb_sink(box.value)
  }
}
